Sources/SourceKitLSP/CapabilityRegistry.swift

```diff
@@ -0,0 +1,103 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import LanguageServerProtocol
+
+/// Handler responsible for registering a capability with the client.
+public typealias RegistrationHandler = (_ registration: CapabilityRegistration) -> Void
+
+/// A class which tracks the client's capabilities as well as our dynamic
+/// capability registrations in order to avoid registering conflicting
+/// capabilities.
+public final class CapabilityRegistry {
+  /// Registered completion options.
+  private var completion: [CapabilityRegistration: CompletionRegistrationOptions] = [:]
+
+  public let clientCapabilities: ClientCapabilities
+
+  public init(clientCapabilities: ClientCapabilities) {
+    self.clientCapabilities = clientCapabilities
+  }
+
+  public var clientHasDynamicCompletionRegistration: Bool {
+    clientCapabilities.textDocument?.completion?.dynamicRegistration == true
+  }
+
+  /// Dynamically register completion capabilities if the client supports it and
+  /// we haven't yet registered any completion capabilities for the given
+  /// languages.
+  public func registerCompletionIfNeeded(
+    options: CompletionOptions,
+    for languages: [Language],
+    handler: RegistrationHandler
```

Comment:
for "handler", how about the more active `registerOnClient`?

Reply:
Done, much better